datafusion/physical-optimizer/src/coalesce_batches.rs

```diff
@@ -57,23 +55,7 @@ impl PhysicalOptimizerRule for CoalesceBatches {
         let target_batch_size = config.execution.batch_size;
         plan.transform_up(|plan| {
             let plan_any = plan.as_any();
-            let wrap_in_coalesce = plan_any
-                // Don't need to add CoalesceBatchesExec after a round robin RepartitionExec
-                .downcast_ref::<RepartitionExec>()
-                .map(|repart_exec| {
-                    !matches!(
-                        repart_exec.partitioning().clone(),
-                        Partitioning::RoundRobinBatch(_)
-                    )
-                })
-                .unwrap_or(false);
-
-            if wrap_in_coalesce {
-                Ok(Transformed::yes(Arc::new(CoalesceBatchesExec::new(
-                    plan,
-                    target_batch_size,
-                ))))
-            } else if let Some(async_exec) = plan_any.downcast_ref::<AsyncFuncExec>() {
+            if let Some(async_exec) = plan_any.downcast_ref::<AsyncFuncExec>() {
```

Comment:
looks like the only thing left is AsyncFuncExec and we could entirely remove the CoalesceBatches pass (and maybe even structure)